OpenRA.Mods.Common/Traits/DeliversCash.cs

```diff
@@ -65,49 +67,64 @@ public string VoicePhraseForOrder(Actor self, Order order)
 				return null;
 
-			return info.Voice;
+			return Info.Voice;
 		}
 
 		public void ResolveOrder(Actor self, Order order)
 		{
 			if (order.OrderString != "DeliverCash")
 				return;
 
-			self.QueueActivity(order.Queued, new DonateCash(self, order.Target, info.Payload, info.PlayerExperience));
+			self.QueueActivity(order.Queued, new DonateCash(self, order.Target, this));
 			self.ShowTargetLines();
 		}
 
 		void INotifyCashTransfer.OnAcceptingCash(Actor self, Actor donor) { }
 
 		void INotifyCashTransfer.OnDeliveringCash(Actor self, Actor acceptor)
 		{
-			if (info.Sounds.Length > 0)
-				Game.Sound.Play(SoundType.World, info.Sounds, self.World, self.CenterPosition);
+			if (Info.Sounds.Length > 0)
+				Game.Sound.Play(SoundType.World, Info.Sounds, self.World, self.CenterPosition);
 		}
 
 		public class DeliversCashOrderTargeter : UnitOrderTargeter
 		{
-			public DeliversCashOrderTargeter()
-				: base("DeliverCash", 5, "enter", false, true) { }
+			DeliversCashInfo info;
+
+			public DeliversCashOrderTargeter(DeliversCashInfo info)
+				: base("DeliverCash", 5, "enter", false, true)
+			{
+				this.info = info;
+			}
 
 			public override bool CanTargetActor(Actor self, Actor target, TargetModifiers modifiers, ref string cursor)
 			{
-				var type = self.Info.TraitInfo<DeliversCashInfo>().Type;
-				var targetInfo = target.Info.TraitInfoOrDefault<AcceptsDeliveredCashInfo>();
-				return targetInfo != null
-					&& targetInfo.ValidStances.HasStance(target.Owner.Stances[self.Owner])
-					&& (targetInfo.ValidTypes.Count == 0
-						|| (!string.IsNullOrEmpty(type) && targetInfo.ValidTypes.Contains(type)));
+				var adc = target.TraitOrDefault<AcceptsDeliveredCash>();
+				if (adc == null)
+					return false;
+
+				if (adc.IsTraitDisabled)
+					return false;
+
+				var type = info.Type;
+				return adc.Info.ValidStances.HasStance(target.Owner.Stances[self.Owner])
+					&& (adc.Info.ValidTypes.Count == 0
+						|| (!string.IsNullOrEmpty(type) && adc.Info.ValidTypes.Contains(type)));
 			}
 
 			public override bool CanTargetFrozenActor(Actor self, FrozenActor target, TargetModifiers modifiers, ref string cursor)
 			{
-				var type = self.Info.TraitInfo<DeliversCashInfo>().Type;
-				var targetInfo = target.Info.TraitInfoOrDefault<AcceptsDeliveredCashInfo>();
-				return targetInfo != null
-					&& targetInfo.ValidStances.HasStance(target.Owner.Stances[self.Owner])
-					&& (targetInfo.ValidTypes.Count == 0
-						|| (!string.IsNullOrEmpty(type) && targetInfo.ValidTypes.Contains(type)));
+				var adc = target.Info.TraitInfoOrDefault<AcceptsDeliveredCashInfo>();
+				if (adc == null)
+					return false;
+
+				// TODO: FrozenActors don't yet have a way of caching conditions, so we can't filter disabled traits
+				// This therefore assumes that all AcceptsDeliveredCash traits are enabled, which is probably wrong.
+				// Actors with FrozenUnderFog should therefore not disable the AcceptsDeliveredCash trait.
```

Comment:
Uh, aren't actors with FrozenUnderFog the *only* things that have the `AcceptsDeliveredCash`, and therefore can disable it?

Reply:
I copy pasted the comment from Capturable:. Yes, ADC is on buildings which has FUF, but they can only enter ally buildings so the said building will never actually be frozen. Maybe the comment should be removed, i'm not sure.

Reply:
Good point. Can you please amend the comment here and in `DeliversExperience` to add "if ValidStances includes Enemy actors"?